### ISSUE TITLE
Comments

### ISSUE THREAD

**Issue by BookOwl** [BookOwl/haumea#6]
This is probably important. 😉 
EDIT:
After discussion, I've decided to use `/* Comment */`

**Comment by BookOwl** [comment]
# What should the comment syntax be?

1. `/* C style  */`
2. `{ Pascal style }`
3. `# Python-ish style #`
4. `<| Something I made up |>`
5. `{| Something else I made up |}`
6. `<- HTML-ish ->`
7. `(| Bananas |)`
8. `(: Emojis :)`

@liam4 @nanalan @jonathan50 @TheMonsterFromTheDeep @algmwc5 please tell me which one you guys think is best (and be serious please)

**Comment by joker314** [comment]
I am not one of the people listed, but oh well. I'll rank them in order, so my favourite will be at the top, and the one I dislike most will be 8th

1. 6
2. 1
3. 2
4. 8
5. 3
6. 4
7. 7
8. 5

5 just seems to be vertically misaligned and thus displeasing to the eye.

**Comment by jupiterio** [comment]
I would prefer C style comments as they're used in a good amount of languages and sometimes pseudocode. Line comments (in the form of  `//`) would be nice too instead of only multiline.

**Comment by TheMonsterFromTheDeep** [comment]
I like both C-style and Python-ish style. I'm honestly actually not sure which one I like better - although I think whichever one you choose should support block comments.

**Comment by joker314** [comment]
I like many of the brackets where the opening and closing character is different. Oh, and an interesting feature I would suggest is allow you to escape comment endings. Not sure how useful this would be, but a
```
< My comment \> continues till here> and then stops
```

might be cool, not sure though.

I don't like the Python-ish one much because hashtags have a rightward slant that makes my brain think of it as an opening "bracket". If there was a hashtag in ASCII that slants in the opposite direction I'd say that could be a pretty cool comment system.

**Comment by BookOwl** [comment]
@joker314, I just didn't know your Github account.

Right now I'm leaning towards 1, 2, and 3. Thanks for the input guys! :D

**Comment by TheMonsterFromTheDeep** [comment]
I personally don't really find angle brackets to be that attractive for commenting. However, that may simply be because I'm so used to other symbols for it... :package: 

I think however that the best choice for this language is something that clearly marks said comment as descriptive, and I honestly have no idea what the best syntax to do that might be. I think something with quotation marks might work, like `<" comment >` because that sort of marks it as a note, but then there is the confusion between that and string syntax...

The other symbol I think might work with commenting is the question mark, so maybe something like `<? comment >` would be good. I like the question mark in comments because it sort of looks distinct from code and it also sort of conveys the notion that it's there to help.

I also agree with io that inline comments would be nice too - I think question marks would also work well for that, e.g.
```
? This method calculates the factorial of parameter n.
? This is not the most efficient way to do it, but it works for demonstration purposes.
to factorial with (n) do ...
```

**Comment by Jonathan50** [comment]
Hmmm. I like 2-5.

@TheMonsterFromTheDeep Then you can't name procedures or variables stuff like `is_integer?`. Oh, Haumea doesn't allow question marks in names anyway lol ![](https://cdn.scratch.mit.edu/scratchr2/static/djangobb_forum/img/smilies/tongue.png)

**Comment by jupiterio** [comment]
@Jonathan50 You can't on quite a few other languages either 📦

**Comment by TheMonsterFromTheDeep** [comment]
@Jonathan50 Yeah, there are certainly problems with question marks. :P

I think it's honesty pretty difficult to design a perfect comment syntax for this language, because comment syntax needs to be concise and not interfere with other syntax, but in this case it should also be relatively readable. :/

**Comment by jupiterio** [comment]
@TheMonsterFromTheDeep Maybe `#` or even `--`? Python, Elixir use the former, Lua the latter, and they all are similar to Haumea in style. I'm kinda leaning to them, since, now that I think of it, `\\` feels too "advanced" for a language of this kind.

**Comment by Jonathan50** [comment]
There's the Smalltalk-style `"..."`.

**Comment by algmwc5** [comment]
Shell style 
`#Comments`

**Comment by tjvr** [comment]
9. Files are a superset of markdown. Executable code is indented; comments are everything else...

**Comment by liam4** [comment]
I don't dislike `--`.

**Comment by CosmicWebServices** [comment]
C style

**Comment by BookOwl** [comment]
> 9\. Files are a superset of markdown. Executable code is indented; comments are everything else... 

That's interesting, I actually have heard of a language like that, but I don't know how many people are going to be writing documents that have Haumea code in it.

**Comment by tjvr** [comment]
> I actually have heard of a language like that

I did shamelessly steal the idea from [Eve](https://github.com/witheve/Eve) :)

**Comment by nanalan** [comment]
I like `--`, lua-style, or `?`, both of which continue until newline.
Multi-line comments are needed though.

_If I had to pick one of the 6, I'd say JS/Python style._

**Comment by joker314** [comment]
Or you could do (for begginer friendliness)
```
comment 

end
```

**Comment by CosmicWebServices** [comment]
^^^

**Comment by liam4** [comment]
@joker314 I honestly can't tell if that's sarcastic or not.

**Comment by BookOwl** [comment]
I really like the question marks so far, but what about for multiline comments?

**Comment by liam4** [comment]
```
???
  Blah
???
```

**Comment by BookOwl** [comment]
That could work.

**Comment by CosmicWebServices** [comment]
or 
```
~~~

~~~
```

**Comment by joker314** [comment]
Unless somebody writes a comment like
`WHYY????` which will unfortunately close the comment

**Comment by CosmicWebServices** [comment]
or 
```
***
Comment
***
```

**Comment by liam4** [comment]
`???` could require a line dedicated to itself.

**Comment by nanalan** [comment]
^

**Comment by nanalan** [comment]
```
? I'm a comment.
blah blah do
  ???
    this is a multiline comment
    really??? yes
  ???
  stuff
end

**Comment by CosmicWebServices** [comment]
or

```
***
Comment
***
```

**Comment by liam4** [comment]
[I saw that.](https://github.com/BookOwl/haumea/issues/6#issuecomment-256929065) :P

**Comment by nanalan** [comment]
Provided somebody writes syntax definitions for Haumea, the actually character shouldn't matter too much since comments would be highlighted :wink:

**Comment by joker314** [comment]
In the `***` system, you can't posssibly write `Terms and Conditions apply***` so it should be fine, and allow you to write comments in between tokens:

`while ***this will loop*** x < 5 do`

**Comment by joker314** [comment]
nanalan, yeah, but what if somebody is using Notepad to write Haumea

**Comment by liam4** [comment]
I guess inline comments are nice..?

```
? This will loop
while x < 5 do
```

But I prefer this placement of comments..

**Comment by joker314** [comment]
Of course but some people might write long one liners

**Comment by BookOwl** [comment]
@joker314 THEN THEY ARE NOT WORTHY TO USE HAUMEA!

**Comment by joker314** [comment]
eheehe

**Comment by nanalan** [comment]
```
some stuff ? hi
```

**Comment by liam4** [comment]
```
while (
  this and ? ..
  that and
  this or even that ? ..
)
```

**Comment by BookOwl** [comment]
I really like `? this is a comment` for single line comments, and 
```
???
Hi!
???
``` for multiline comments

**Comment by nanalan** [comment]
gj with the markdown there :sunglasses:

**Comment by liam4** [comment]
I suppose I'm not allowed to +1 that, given I brought up the idea. :tada:

**Comment by nanalan** [comment]
@BookOwl I also like the fact you used the same font that Gwiddle uses (?) for the Haumea logo :shipit:

**Comment by BookOwl** [comment]
Really? I never knew that.

**Comment by BookOwl** [comment]
Also, if users REALLY want inline multiline comments, they could do
```
while ??? This is my condition ??? foo() ??? This is my statement. ??? do
 bar()
end

**Comment by joker314** [comment]
Exactly.

**Comment by BookOwl** [comment]
If anybody has any major objections to question mark comments, speak up now.

**Comment by CosmicWebServices** [comment]
My idea works too 

```
while ***This is my condition*** foo() *** This is my statement. *** do
   bar()
end
```

**Comment by joker314** [comment]
Hey, why not have two comment thingies

**Comment by CosmicWebServices** [comment]
I think I like `***` better than `???` but is just my opinion

**Comment by BookOwl** [comment]
@joker314 That could get confusing having two different comment syntaxes.

**Comment by joker314** [comment]
`***` works when interlacing inside of statements because it is perfectly symetrical.

**Comment by BookOwl** [comment]
That's true

**Comment by nanalan** [comment]
> Also, if users REALLY want inline multiline comments, they could do
>
>     while ??? This is my condition ??? foo() ??? This is my statement. ??? do
>      bar()
>     end
I dislike that - it could create problems with `???` within the comment themselves.

**Comment by BookOwl** [comment]
Great job with markdown @nanalan 😎

**Comment by joker314** [comment]
I was just about to say that

**Comment by CosmicWebServices** [comment]
@BookOwl What will you choose? `***` or `???` I will leave the decision to you. 😎

**Comment by joker314** [comment]
I gtg, 🚆 ride!

**Comment by CosmicWebServices** [comment]
bye

**Comment by BookOwl** [comment]
>> Also, if users REALLY want inline multiline comments, they could do
>>
>>     while ??? This is my condition ??? foo() ??? This is my statement. ??? do
>>     bar()
>>     end

> I dislike that - it could create problems with `???` within the comment themselves.

That's true, I think I'll stick with @liam4's idea of having the `???` have to appear on it's own line.

**Comment by CosmicWebServices** [comment]
Or
`***`
Just saying much simpler

**Comment by BookOwl** [comment]
But that still has the problem with `***` appearing inside a comment.

**Comment by CosmicWebServices** [comment]
Really???

**Comment by BookOwl** [comment]
`while foo() *** Hey, a comment *** <- See that? *** bar()`

**Comment by CosmicWebServices** [comment]
Well this issue exists in a lot of languages it can be avoided less with my version

**Comment by birdoftheday** [comment]
You should probably choose the simplest choice. There's no need to make up something new and complicated just because it's original.

`//` or even `#` would work just fine.

**Comment by birdoftheday** [comment]
That said, I would not be opposed to something original, it would just have to follow these guidelines (for me, at least):

* Nothing too fancy or hard to type like `(: sample :)`
* Nothing too hard to identify as a comment like indentation
* Nothing that could be seen as code (like `?`)

**Comment by nanalan** [comment]
> // or even # would work just fine.
I disagree @birdoftheday, what about **float division**? (`/` vs `//`)

**Comment by nanalan** [comment]
@birdoftheday provided `?` isn't part of the syntax, how could it be seen as code? ;P

**Comment by birdoftheday** [comment]
`condition ? something : somethingelse` is in a lot of languages.

**Comment by birdoftheday** [comment]
It just looks kind of odd to me.
```
? this does something
? it also does something else
5 + 5 ? this calculates five plus five
```

**Comment by birdoftheday** [comment]
It's kind of hard to separate the `5 + 5`, the code ,and the `? this calculates five plus five`, the comment

**Comment by nanalan** [comment]
Not in Haumea :shipit:

Regarding inline ifs, @BookOwl have you considered newlines being optional? It's a functional piece of syntax but it looks quite nice

```
x = if 1 = 1 do 25 end else do 2
```

**Comment by BookOwl** [comment]
`condition ? something : somethingelse` has always been confusing to me, so if I add it it would have different syntax.

**Comment by birdoftheday** [comment]
All I'm saying is that `?` kind of looks like an operator to me. It might be different for other people.

**Comment by birdoftheday** [comment]
@BookOwl Why not just `#`?

**Comment by nanalan** [comment]
@birdoftheday not saying I use them, but have you _ever used an Apple keyboard_? :stuck_out_tongue_winking_eye:

**Comment by birdoftheday** [comment]
@nanalan Yes

**Comment by liam4** [comment]
> not saying I use them, but have you ever used an Apple keyboard? 😜 # is really difficult to type on them :(

pinky on (left) shift, index on 3, ###, :tada:

I really don't find that difficult to type!

**Comment by BookOwl** [comment]
What about @TheMonsterFromTheDeep's idea of using `<? ... ?>`?

**Comment by BookOwl** [comment]
Well, I've programmed skipping comments into the scanner, so as soon as we can decide which comment syntax to use I can push it.

**Comment by BookOwl** [comment]
# FINAL VOTE
1. `/* This is a comment */`
2. `<? This is a comment ?>`
3. `?? This is a comment ??`
4. `-- This is a comment --`
5.  `<- This is a comment ->`

**Comment by BookOwl** [comment]
Currently my favorites are 2 and 3, but I'm open to any of them.

**Comment by nanalan** [comment]
3 and 4.

**Comment by CosmicWebServices** [comment]
1 and 5

**Comment by BookOwl** [comment]
Because I'm busy.

**Comment by jupiterio** [comment]
Every one of them except for 2 and 5. I don't really imagine <> being used in a language of this style, other than for comparing.

**Comment by BookOwl** [comment]
What do you think of 3?

**Comment by nanalan** [comment]
> What do you think of 3?

It's nice, but I like the idea of having different start/end tokens.

**Comment by joker314** [comment]
There are 6 reaction emojis, so why not vote using them

1 - +1
2 - -1
3 - Laugh
4 - Hooray
5 - Confused

Please select with the option below. I have initialised them all so that you can just click the one you want. However, this means I cannot vote. You ARE allowed to vote more than once.

**Comment by tjvr** [comment]
> pinky on (left) shift, index on 3, ###, 🎉

Alt+3 on non-US Mac keyboards. It did take me a while to get used to it when I first switched from a British PC layout. Not a big deal now though...

**Comment by Firedrake969** [comment]
`?? This is a comment ??`

Are you sure?  :P

**Comment by nanalan** [comment]
1. `/* comment */` it is, then?

**Comment by BookOwl** [comment]
@nanalan I think so.

**Comment by BookOwl** [comment]
Comments have been added to master with commit 0db0ee9 

They use the `/* I'm a comment */` syntax. Thanks for the help everyone! :D

**Comment by Jonathan50** [comment]
> Multi-line comments are needed though.

Why do you think that?

> @joker314 I honestly can't tell if that's sarcastic or not.

Why do you think that?

> I saw that. :P

Your link is broken.

> gj with the markdown there :sunglasses: 

How can you see that&hellip;?

**Comment by nanalan** [comment]
>> I saw that. :P
> Your link is broken.

His old comment, which he deleted.

>> gj with the markdown there :sunglasses:
> How can you see that…?

Failed `>` quote :stuck_out_tongue_winking_eye:

**Comment by liam4** [comment]
> > @joker314 I honestly can't tell if that's sarcastic or not.
>
> Why do you think that?

@Jonathan50 I've found everything *far* too verbose in (at least some) languages like this. Having a huge comment syntax will just distract from the content of the comment! It's not particularly friendly to beginners - it's obvious, but once you've learned (the syntax) it's not hard to remember.

And I figured joker314 wasn't being serious when he suggested that (additionally, since it's a pain to type!).
